CMakeLists.txt

```diff
@@ -12,12 +12,26 @@ endif()
 
 set(CMAKE_CXX_FLAGS_RELEASE "-O2 -DNDEBUG")
 
-set(ONNX_MLIR_SRC_ROOT "${CMAKE_CURRENT_SOURCE_DIR}")
-set(ONNX_MLIR_BIN_ROOT "${CMAKE_CURRENT_BINARY_DIR}")
+# CMAKE_CFG_INTDIR is . for single-config generators such as make, and
+# it has a value (e.g. $(Configuration)) otherwise, so we can use it to
+# determine whether we are dealing with a multi-config generator.
+if (NOT "${CMAKE_CFG_INTDIR}" STREQUAL ".")
+  set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib/$<CONFIG>)
+  set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib/$<CONFIG>)
+  set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin/$<CONFIG>)
+else()
+  set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+  set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+  set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+endif()
+
```

Comment:
Do we even need to check for "." here? Since "." is basically a noop in a path, we can just set everything to `.../$<CONFIG>` (and indicate such in the comments)?

Reply:
Yes, we do. CMAKE_CFG_INT_DIR is . with single-config generators, but $<CONFIG> always expands to all configurations. So even with make or ninja, the output directories would contain Debug/Release/etc. in the path. My assumption is that we would rather not do that and maintain the current pattern of having build_dir/bin and build_dir/lib when using ninja or make rather than build_dir/bin/release and build_dir/lib/release.

Reply:
I actually don't see what's wrong with having Debug/Release/etc for make or ninja. At least we would then be consistent across the board. But maybe a better path is `${CMAKE_BINARY_DIR}/$<CONFIG>/{bin|lib}`?

Reply:
LLVM uses build_dir/{bin|lib} for ninja/make and build_dir/build_flavor/{bin|lib} for Xcode/Visual Studio. It is definitely more convenient as a user as well as for scripts and tools to have the build flavor as a common parent for bin and lib, so I was going to suggest that in a separate code review - since making the change is not required for the multi-config generators to work. I can make it as part of this pull request though if you prefer.

As for having the build flavor for make or ninja as well - that is generally not how projects that use them are set up, but it is certainly convenient if anyone is going to build more than one flavor. I was trying to maintain the setup as it was for those generators, but I can change it if you would prefer.

Reply:
OK up to you whether you want to make the change for the multi-config generators in this PR or another one.

@tjingrant @caoimhinuibrian I vote for having Debug/Release/... etc for make/ninja as well. It would make it easier to compare different flavors if needed. And it also makes the cmake file logic simpler and consistent across the board. Cast your vote and we can go with the majority.

Reply:
I'm fine with doing that

Reply:
For what it's worth, the behavior of prepending the Config for multi-config generators, but not single-config, is baked into CMake, as [described here](https://cmake.org/cmake/help/latest/prop_tgt/RUNTIME_OUTPUT_DIRECTORY.html). That doesn't mean ONNX MLIR can't decide to override the default here, but it would be non-standard.

Reply:
Seeing that we are actually having issues with Debug vs Release build (so we likely need to compare them), and with Kevin's vote, let's go with having build flavor for make/ninja as well, i.e., same for both single and multi config across the board. Thanks.

Reply:
I made the change to add the build flavor to the single-config generators. I also switched where the build flavor is in the path to build_dir/build_type/{bin|lib}. I would have preferred to do that in a separate change, but that would require adding support for having bin/build_type and lib/build_type to MainUtils.cpp which seems unnecessary if the plan is to just switch the order later.